test/functional/lua/buffer_updates_spec.lua

```diff
@@ -305,6 +305,19 @@ describe('lua buffer event callbacks: on_lines', function()
     n.assert_alive()
   end)
 
+  it('no invalid lnum error for closed memline in on_detach #31251', function()
+    exec_lua([[
+      vim.api.nvim_buf_set_lines(0, 0, -1, false, { '' })
+      local bufname = 'buf2'
+      local buf = vim.api.nvim_create_buf(false, true)
+      vim.api.nvim_buf_set_name(buf, bufname)
+      vim.bo[buf].bufhidden = 'wipe'
+      vim.cmd('vertical diffsplit '..bufname)
+      vim.api.nvim_buf_attach(0, false, { on_detach = function() vim.cmd("redraw") end})
+      vim.cmd.quit()
```

Comment:
usually `n.expect_exit()` is needed when actively quitting the current test Nvim instance. should we do that here?

Reply:
Oh. Then `vim.cmd.close()` would be clearer.

Reply:
The Nvim instance isn't closed here, just one of two split windows.

Reply:
Does `vim.cmd.close()` not have the required behavior?

Reply:
Yes that also works, `:close` being the same as `:quit` except when closing the last window. But actually we are testing deleting the buffer, so I'll update it to `:bdelete`.